### ISSUE TITLE
Question: how is the "only" attribute of a schema intended to be used?

### ISSUE THREAD

**Issue by justin-richert** [marshmallow-code/marshmallow#761]
From the marshmallow docs:

> only (tuple) – A list or tuple of fields to serialize. If None, all fields will be serialized. Nested fields can be represented with dot delimiters.

This gives me the impression that it is a separate attribute from fields that is used on serialization to know which fields to serialize. This is partly true in that if `only` is set on a schema instance and dump is called, it does indeed affect which fields are serialized. But if you look at the schema instance after dump is called, it actually has "destructively" modified the `fields` attribute of the schema. From my perspective, it seems as though `fields` should remain unchanged -- `only` should merely inform the serialization process during the execution of the serialization as opposed to altering the schema instance. I am possibly misunderstanding how it was intended to be used, but the docs may need to be clarified in that case.

A little example to further explain my use case and why I hoped it would behave the way I imagined (using marshmallow with Flask):

```python
from schemas import SchemaClass

@app.route('/collection/<resource_id>')
def route_handler(**kwargs):
    return getattr(ViewHelperClass, request.method.lower())(**kwargs)


class ViewHelperClass:
    @staticmethod
    @SchemaClass(many=False, exclude=['field1'], dump_only=['field2', 'field3'])
    def get(**kwargs):
       # Do some stuff
       # return something
```

In my case, I've provided a base schema which all other schemas (including `SchemaClass` in the above example) that extends the functionality a bit to allow an instance of a schema to decorate a function. This is highly useful, because I know all the attributes of the schema instance prior to execution time.

However, I allow the client to request only certain fields be returned: `api.example.com/collection/1?fields=field3,field4`. The only way to satisfy that request is by setting the `only` attribute of the schema. This has some unfortunate side effects, though, because the schema instance is not recycled for the next request, meaning the fields have been permanently altered for future requests when `dump` is called with a subset of fields supplied to `only`.

Sorry for the long explanation, but hopefully that helps clarify the issue I'm having. To reiterate my question: was `only` to be used exclusively on one-off instances of schemas? If that is the case and to remain the case, would it be possible to provide an argument for serialization that allows specifying this? For example, `schema_inst.dump(data, only=only_the_fields_I_want)`?

**Comment by scottwernervt** [comment]
This feature would be very useful for web applications, along with `exclude` argument:  `schema.dump(data, exclude=(field1, ...))`.

**Comment by deckar01** [comment]
I'm not exactly sure what your use case is. I do not understand the way you are using `@SchemaClass`. I usually declare my schemas statically and use them in routes as needed. In your case you would probably provide a schema instance to your view helper and reuse it throughout the life of the view helper instance.

Marshmallow does require you to construct a new schema instance to modify the field set. It is a convenience to avoid duplicating schema declaration logic that is a subset of another schema. If you need to dynamically modify the schema fields per method call, you need to construct a new schema instance. Currently the schema caches it's customized configuration at construction time and it would make things a lot more complicated to invalidate that configuration in methods like dump.

Have you looked at webargs? https://github.com/sloria/webargs

It is a very nice framework for validating API arguments.

**Comment by justin-richert** [comment]
Thanks for your reply!

I think I may need to clarify the explanation of my use case just a little bit.

In using it as a decorator, it statically declares the schema and will never be declared again for the entire lifetime of the Flask app. So I definitely agree with your approach of statically declaring schemas.

The problem I'm having is I **don't** want to modify the field set on the schema instance, but setting the `only` attribute of the schema instance and calling `schema_instance.dump` **does** modify the schema instance. Inspecting `schema_instance.fields` before and after serializing with `only` set to a subset of `schema_instance.fields` reveals this.

I don't necessarily have a major issue with the above behavior (though it is a little confusing) because I'm declaring the schema statically, so I'd really rather not have to set `schema_instance.only` on every request (kinda violates the static nature of it). But I can't find an alternative way to specify which fields of the schema I want to serialize *at serialization time*. The documentation seems to indicate `schema_instance.only` is the way to specify it, but I would prefer `schema_instance.dump(data, only=only_the_fields_I_want_serialized)`.

**Comment by justin-richert** [comment]
Oh, I forgot to mention - I looked into webargs. While it doesn't really solve my current issue, it does look to be a great framework. Thanks for the recommendation!

**Comment by lafrech** [comment]
I might be missing something, but can't you

- declare a Schema class in your decorator and keep it for the entire lifetime of the app
- instantiate the Schema with `only` before dumping

?

The class wouldn't be modified.

**Comment by justin-richert** [comment]
I'm actually *instantiating* a schema at decoration time as opposed to just declaring the class. So the schema instance is around for the lifetime of the app. Which is actually really useful to me because I don't really need to alter anything on the schema except this pesky `only` attribute :)

**Comment by lafrech** [comment]
Do you _have_ to instantiate in the decorator? Couldn't you instantiate just before the dump? Is it just to avoid the cost of instantiation?

**Comment by justin-richert** [comment]
No, I definitely don't have to. And that may be the resolution to this thread, and that's ok. I just thought it a conversation worth having/question worth asking, because it seemed like it might prove to be a useful feature. I will alter my approach, at least for the time being!

To answer the why behind it: I also use it to load/for validation at the beginning of the request. It cleaned my code up quite a bit to be able to decorate with an instance instead of piping the initialization variable through with another decorator, e.g. `request_schema(schema_cls, many=False, exclude=....)`

**Comment by lafrech** [comment]
No problem, you're welcome. I was just trying to understand. Glad you're not stuck.

**Comment by justin-richert** [comment]
Just out of curiosity -- if I put together a PR to implement the functionality I'm looking for, would that be considered, or is this functionality something that would actually be opposed?

Either is fine. I would love to contribute if it would be helpful to others, but I of course also don't want to appear as trying to force functionality that is overall NOT desired by others.

**Comment by deckar01** [comment]
Are you proposing adding an `only` argument to `dump()`?

Here is how it currently works:

```py
from marshmallow import fields, Schema

class Example(Schema):
    id = fields.Integer()
    name = fields.String()

data = {'id': 1, 'name': 'test'}

example = Example(only=('name',))
assert example.fields.keys() == ['name']

result = example.dump(data)
assert result.data == {'name': 'test'}
assert example.fields.keys() == ['name']
```

Can you copy and change this example to show me what you are doing differently and how you would expect it to behave?

**Comment by justin-richert** [comment]
Yeah, adding an `only` parameter to dump to make a schema instance fully static while having the ability to serialize a subset of fields.

```python
from marshmallow import fields, Schema

class Example(Schema):
    id = fields.Integer()
    name = fields.String()

data = {'id': 1, 'name': 'test'}

example = Example()
assert example.fields.keys() == ['id', 'name']

result = example.dump(data, only=['name'])
assert result.data == {'name': 'test'}
assert example.fields.keys() == ['id', 'name']
```

The reason for this is that the schema instance could be reused if this functionality existed, and my application provides the ability to request a subset of fields for a resource on a `GET` request. The schema should still have the full field set as it is used for to load and dump on future requests that may want ALL fields (or a different subset of fields) but inherently still has all the same deserialization and serialization rules.

But I just wanted to feel out if this was something that might be helpful to others and whether or not the PR might be considered before spending some time on it. As I know I am asking for something that may not be at the top of a priority list, I wondered if it might be accepted as a contribution instead.

If not, I promise I'll move on and we can close the issue :)

**Comment by deckar01** [comment]
I am hesitant to support another way to do the same thing. Is this really a performance optimization? Are you sure that flask doesn't cause the schema to be instantiated on each request anyway?

**Comment by justin-richert** [comment]
I'm not sure that it's a significant performance optimization. Obviously instantiation cost is saved, but I understand that isn't a major motivator. Because I saw the duplication in my code and saw a way to solve it, I jumped at the chance to cut down to a single instantiation at decoration -- and it works really nicely other than this issue. I did ensure that flask does not cause the schema to be instantiated on each request.

A good chunk of the benefit to me was also in the code organization. It allowed me to abstract/organize quite a bit better when decorating with an instance.

I appreciate the time you guys took to think through this with me! And thanks for such a great project.
